api/envoy/extensions/load_balancing_policies/deterministic_aperture/v3/deterministic_aperture.proto

```diff
@@ -0,0 +1,36 @@
+syntax = "proto3";
+
+package envoy.extensions.load_balancing_policies.deterministic_aperture.v3;
+
+import "envoy/extensions/load_balancing_policies/ring_hash/v3/ring_hash.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.load_balancing_policies.deterministic_aperture.v3";
+option java_outer_classname = "DeterministicApertureProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/deterministic_aperture/v3;deterministic_aperturev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Deterministic Aperture Load Balancing Policy]
+// [#not-implemented-hide:]
+
+// Deterministic Aperture load balancing is designed to conserve connections between Envoy and
+// backend instances. This is achieved by dividing the backends uniformly among participating peers.
+// This way, we avoid a complete mesh of connections between Envoy and backends. The participating
+// Envoy peers are placed on a ring that is conceptually laid on top of the Ring Hash of backends
+// to determine their overlaps. By using the ring overlaps as a mechanism to divide the backends,
+// the backends get divided as a fraction of their overlaps with each peer. The algorithm uses a
+// uniform random distribution to select a backend from the range of overlapping backends. This
+// along with P2C ensures uniform load distribution.
+// [#extension: envoy.load_balancing_policies]
+message DeterministicApertureLbConfig {
+  // Hash ring configuration for the internal ring for endpoints.
+  ring_hash.v3.RingHash ring_config = 1;
```

Comment:
Is this new policy ring-hash-specific?  If so, maybe instead of implementing this as a separate policy, it should just be a new config field in the existing ring-hash policy?

That having been said, I think a better approach would be to change this such that it is not actually specific to the ring_hash policy in the first place.  I haven't looked at the implementation to see how tightly coupled this is with ring_hash, but if the overall goal here is to allow the LB policy to choose from only a subset of the endpoints, then presumably the child policy here could be any policy, even something other than ring_hash.  In that case, this field could be replaced with:

```
config.cluster.v3.LoadBalancingPolicy child_policy = 1;
```

Reply:
But why is the implementation of this new subsetting policy dependent upon implementation details of the child policy in the first place?  It seems to me that this policy should be self-contained: it should determine which subset of hosts the child policy should be looking at and then delegate to the child policy to pick from that set of hosts.  It should not care how the child policy is actually doing that pick.

If we agree that this is the long-term approach that we want here, then we should definitely not bake any dependency on ring_hash into the API itself.  So I would definitely prefer to see a generic child policy config here.

I'll defer to @wbpcode in terms of the implementation, but in principle, it seems wrong to me to implement this in a way that works only with ring_hash, regardless of the API representation.  It seems like this is just unnecessarily breaking abstractions by exposing inner workings of the ring_hash policy, and it is unnecessarily constraining to provide this functionality only in conjunction with the ring_hash policy.

I would really prefer to see this implemented in a way that is completely agnostic to the details of the underlying child policy.

Reply:
I think this might need some explanation :).  The `deterministic aperture` lb uses a `ring` to place its endpoints but then overlays the Envoy peers ring on top to split the endpoints `equally` among the peers. This is well explained here : https://blog.twitter.com/engineering/en_us/topics/infrastructure/2019/daperture-load-balancer.

Reply:
I understand that the algorithm here is using a ring, and I agree that it makes sense to reuse the existing ring implementation in Envoy.  The thing I'm objecting to is the way we're accomplishing that reuse.  This PR is currently doing that by tying the deterministic_aperature policy to the ring_hash policy, which (a) prevents flexibility by not allowing deterministic_aperature to be used with other child policies, even though in principle its functionality seems orthogonal to that of the child policy, and (b) breaks the abstraction of the ring_hash policy by tying another policy to its internals, which will make it harder to make changes to ring_hash in the future.

I think a cleaner way to structure this would be to refactor the ring implementation out of the ring_hash policy and into its own library, and then use that ring library in both the ring_hash policy and in the deterministic_aperature policy.  That way, you can still reuse the existing ring implementation, but nothing in the deterministic_aperature policy will have to reach inside of the ring_hash policy, and it will allow using deterministic_aperature with any child policy.

Reply:
I think it is a good suggestion - refactoring the `ring` implementation to a library which can be reused. I am not sure if we want to represent the internal structure as an abstraction , atleast in its initial implementation. IMHO although theoretically it is a good idea, i think the extra complexity might not be used much. Let me know what you think.

Thanks a lot for your input!

Reply:
Also, I think the Ring configuration for Deterministic Aperture would look the same as for Ring Hash balancer's Ring. So not sure how to structure this.

Reply:
> I am not sure if we want to represent the internal structure as an abstraction

I'm not sure what you mean by that.  I think the ring can be represented using an interface that looks something like this:

```
class Ring {
 public:
  struct RingEntry {
    uint64_t hash;
    Host* host;
  };

  Ring(size_t min_size, size_t max_size, HashFunc hash_func);

  // Returns the size of the ring.
  size_t size() const;

  // Returns the entry at the specified index.
  RingEntry& entry(size_t index);

  // Returns the index to use for the specified hash.
  size_t FindIndex(uint64_t hash);
};
```

> I think the Ring configuration for Deterministic Aperture would look the same as for Ring Hash balancer's Ring

I think it would be fine to just duplicate the relevant config fields from the ring_hash config into the deterministic_aperature config.

Reply:
What I meant was

Reply:
> What I meant was that I am not sure if we need an abstraction of LoadBalancingPolicy to represent the `Ring` as I am not sure if we will be implementing another `LoadBalancingPolicy` for all practical purposes.

Sorry, I don't understand that part.

I agree that the ring should not be represented as a `LoadBalancingPolicy` object, if that's what you mean.  But it sounds like we are already agreeing on the rough shape of the API for the ring object.

I am also not sure what you mean by not implementing another `LoadBalancingPolicy`.  I think the deterministic_aperature LB policy will still need to be a `LoadBalancingPolicy`; it will just be a policy that filters the set of hosts to pass to the child policy and then delegates to the child policy to do the pick.

Reply:
> I  agree that the ring should not be represented as a LoadBalancingPolicy object, if that's what you mean. 

Yes thats what i meant :)

I am a bit concerned that we will have two messages (one each for Deterministic Aperture and Ring Hash load balancers) to represent the same underlying configuration for Ring.  What do you think?

Reply:
I think it's fine to have two instances of the ring config fields, since there will be two ring instances.  There isn't actually any requirement that they need to be configured with the same values.  In fact, given that the two rings will actually not have the same set of hosts in them (the one in the ring_hash child policy will have only the subset of hosts selected by the parent policy), I can imagine cases where users would actually want to tune the ring sizes differently for the two instances.

Reply:
What i meant was that the proto message for both would most likely be same. Since we were discussing having each load balancer having its own Ring message, I was not sure if that is ideal.

Reply:
I think it's fine to just duplicate the necessary fields in both of the LB policy configs.  From the perspective of the configs, the fact that the underlying policies happen to share ring implementations is just an implementation detail, not something that should be baked into the API.

Reply:
Sounds good.  Thanks for the suggestions. 

So in essence, here is what we have concluded:
1. Refactor Ring as a lib so that it could be used by both load balancers
2. Have separate RingConfig for both load balancer (RingHash already has one)

Did i miss anything?

Reply:
I don't think we need a separate `RingConfig` message.  I think it would be fine to just copy the relevant fields from the `RingHash`  message to the new `DeterministicApertureLbConfig` message.

But otherwise, yes, this sounds good!

Reply:
Thanks @markdroth .  Since the changes discussed here implies we change the existing RingHash load balancer to use the new refactored Ring implementation, I am wondering if that will increase the scope of this PR. One way to do this could be to not change the existing RingHash load balancer in this PR but have only DeterministicAperture load balancer use the new the Ring library. Then when we move the RingHash load balancer to the new extension model, change it to use the RIng library. I can give a shot at moving the RingHash load balancer to the extension model in another PR.

Reply:
I'll let @wbpcode weigh in here, but IMHO, refactoring the ring functionality out of the ring_hash LB policy really has nothing to do with converting the ring_hash LB policy to be an extension.  I think if you're going to be moving that functionality into its own library, it makes sense to change ring_hash to use that library in the same PR, rather than just duplicating the functionality and then removing the duplication later.

Reply:
Got it. Thanks for the clarification 👍

Reply:
@markdroth I have refactored out the `Ring` concept.  PTAL.
The `docs` build is still broken unfortunately.